azurerm/resource_arm_monitor_diagnostics.go

```diff
@@ -0,0 +1,284 @@
+	retentionDays := d.Get("retention_days").(int)
+
+	allMetricSettings, allLogSettings, err := getAllDiagnosticSettings(targetResourceId, meta)
+	if err != nil {
+		return err
+	}
+
+	allSettings := append(*allMetricSettings, *allLogSettings...)
+	if !utils.StringSliceContainsStringSlice(disabledSettings, allSettings) {
+		return fmt.Errorf("Invalid value for disabled settings provided, use one or multiple of: %q", allSettings)
+	}
+
+	if len(allSettings) == len(disabledSettings) {
+		return fmt.Errorf("You can not disable all settings, rather delete diagnostic logging")
+	}
+
+	diagnosticSettings := &insights.DiagnosticSettings{}
+	diagnosticSettings.Metrics = expandMetricsConfiguration(*allMetricSettings, disabledSettings, retentionDays)
+	diagnosticSettings.Logs = expandLogConfiguration(*allLogSettings, disabledSettings, retentionDays)
+
+	if len(storageAccountId) > 0 {
+		diagnosticSettings.StorageAccountID = &storageAccountId
+	}
+
+	if len(workspaceId) > 0 {
+		diagnosticSettings.WorkspaceID = &workspaceId
+	}
+
+	if len(eventHubAuthorizationRuleId) > 0 && len(eventHubName) > 0 {
+		diagnosticSettings.EventHubAuthorizationRuleID = &eventHubAuthorizationRuleId
+		diagnosticSettings.EventHubName = &eventHubName
+	}
+
+	_, err = client.CreateOrUpdate(
+		ctx,
+		targetResourceId,
+		insights.DiagnosticSettingsResource{
+			DiagnosticSettings: diagnosticSettings,
+		},
+		name)
+	if err != nil {
+		return fmt.Errorf("Error creating Diagnostics Setting %q (Resource ID %q): %+v", name, targetResourceId, err)
+	}
+
+	read, err := client.Get(ctx, targetResourceId, name)
+	if err != nil {
+		return err
+	}
+	if read.ID == nil {
+		return fmt.Errorf("Cannot read ID for Monitor Diagnostics %q", name)
+	}
+
+	d.SetId(*read.ID)
+
+	return resourceArmMonitorDiagnosticsRead(d, meta)
+}
+
+func resourceArmMonitorDiagnosticsRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).monitorDiagnosticSettingsClient
+	ctx := meta.(*ArmClient).StopContext
+
+	monitoringId := parseMonitorDiagnosticId(d.Id())
+
+	resp, err := client.Get(ctx, monitoringId.ResourceID, monitoringId.Name)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Error making Read request on Diagnostic Setting %s: %+v", monitoringId.Name, err)
+	}
+
+	d.Set("name", *resp.Name)
+
+	// ID of base resource is not returned by API, so we have to guess here
+	monitoringId = parseMonitorDiagnosticId(d.Id())
+	d.Set("target_resource_id", monitoringId.ResourceID)
+
+	if resp.StorageAccountID != nil {
+		d.Set("storage_account_id", *resp.StorageAccountID)
+	}
+
+	if resp.EventHubName != nil {
+		d.Set("event_hub_name", *resp.EventHubName)
+	}
+
+	if resp.EventHubAuthorizationRuleID != nil {
+		d.Set("event_hub_authorization_rule_id", *resp.EventHubAuthorizationRuleID)
+	}
+
+	if resp.WorkspaceID != nil {
+		d.Set("workspace_id", *resp.WorkspaceID)
+	}
```

Comment:
since we don't need to dereference the pointer - we can just set these directly (and Terraform will handle setting empty strings as appropriate)